### PR TITLE
Add torch::hash (C++ hashing utilities) and improve arg checking in JIT

### PR DESCRIPTION
**Pull request by apaszke**
Fixes a regression in arg checking when it was moved to C++. Also, added a test so we won't miss it again.

### CODE REVIEW

File: torch/csrc/utils/hash.h [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+};
+
+// Specialization for std::tuple
+template<typename... Types>
+struct hash<std::tuple<Types...>> {
+  template<std::size_t idx, typename... Ts>
+  struct tuple_hash {
+    std::size_t operator()(const std::tuple<Ts...>& t) const {
+      return hash_combine(_hash_detail::simple_get_hash(std::get<idx>(t)),
+												  tuple_hash<idx-1, Ts...>()(t));
```
Comment (fmassa) [inline torch/csrc/utils/hash.h]: nit: there are a bunch of tabs in here

Comment (apaszke) [reply torch/csrc/utils/hash.h]: Thanks! I fixed it

**Review by zdevito — APPROVED** [review]
Looks good! 
We might want to re-time to make sure checking size info did not add overhead back it. I'm guessing it didn't though.

**Comment by apaszke** [comment]
I didn't notice any significant changes, but will run the benchmarks again to make sure

**Comment by apaszke** [comment]
It seems to add 1-2us in the LSTM benchmarks, I'll look into the overhead for more (e.g. 100) inputs. It can be improved by flattening the vectors for sizes, but I didn't want to optimize prematurely.

**Comment by apaszke** [comment]
Build failure is legit. Don't merge it

**Comment by Coderx7** [comment]
Is this hash usuable with Tensors, and modules as well? 
Are there any tests/examples showing how to use this? becasue my attempts at using it like : 
```cpp
std::cout << "x == x2 (hash)? " << (torch::get_hash<torch::Tensor>(x) == torch::get_hash<torch::Tensor>(x2)) << std::endl;
std::cout << "x == x_clone (hash)? " << (torch::get_hash<torch::Tensor>(x) == torch::get_hash<torch::Tensor>(x_clone)) << std::endl;
```
fails with the following errors which I dont know is a bug or me lacking an understanding here . and there is no explanation in the documentation at all : 
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	C2893	Failed to specialize function template 'unknown-type torch::_hash_detail::dispatch_hash(const T &)'	Detector	D:\libtorch\include\torch\csrc\utils\hash.h	78	
Error	C2672	'dispatch_hash': no matching overloaded function found	Detector D:\libtorch\include\torch\csrc\utils\hash.h	78	
```

**Pull request merged**